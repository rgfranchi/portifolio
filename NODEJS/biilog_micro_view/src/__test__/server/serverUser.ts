import { Response } from "miragejs";
import { IUser } from "../../User/UserData";
import { listGroupMock } from "./serverGroup";

export const userRoutes = (actions: any) => {
  let headers = {};
  actions.get("user/findByCompany/:companyId", (schema: any, request: any) => {
    const companyId = request.params.companyId;

    if (companyId.toString() === "9999") {
      return new Response(404, headers, {
        errors: ["Company Id Not Found"],
      });
    }
    const dataBase = schema.db.users.where({ companyId: companyId });

    if (dataBase.length === 0) {
      return new Response(204, headers, "");
    }
    return new Response(200, headers, dataBase);
  });
  actions.get(
    "user/authentication/:email/:password",
    (schema: any, request: any) => {
      const email = request.params.email;
      // recupera password simples.
      const password = request.params.password;
      if (email === "fake@erro.com") {
        return new Response(404, headers, {
          errors: [
            "Servidor não respondeu a solicitação, Entre em contato com o suporte",
          ],
        });
      }
      if (email === "fake@text.com") {
        return new Response(206, headers, {
          errors: ["Servidor respondeu parcialmente a solicitação."],
        });
      }
      if (email === "reposta@text.com") {
        return new Response(200, headers, "Servidor respondeu com texto.");
      }
      const user = schema.db.users.where({
        email: email,
        password: password,
      })[0];
      if (user === undefined) {
        return new Response(401, headers, "Usuário ou Senha inválido");
      }
      const company = schema.db.companies.find(user.companyId);
      const group = schema.db.groups.find(user.groupId);
      const currentUser = {
        userId: parseInt(user.id),
        company: { id: parseInt(company.id), nome: company.nome },
        group: group,
      };
      return new Response(200, headers, currentUser);
    }
  );

  actions.delete("user/:id", (schema: any, request: any) => {
    const id = request.params.id;
    if (id === "1") {
      return new Response(203, headers, { errors: ["Fake Server Delete"] });
    }
    if (id === "9999") {
      return new Response(400, headers, {
        errors: ["Fake Server Delete Error 400"],
      });
    }
    schema.users.find(id).destroy();
    return new Response(202, headers, { errors: ["Fake Server Delete"] });
  });
  actions.get("user/findById/:id", (schema: any, request: any) => {
    if (request.params.id === "-9999") {
      return new Response(400, headers, {
        errors: ["Fake Server Error 400"],
      });
    }
    if (request.params.id === "-1") {
      return new Response(299, headers, { errors: ["Fake Server Error"] });
    }
    const entity = schema.users.find(request.params.id);

    if (entity === null) {
      return new Response(204, headers);
    }
    entity.attrs.password = "";
    return new Response(200, headers, entity.attrs);
    // return schema.operators.find(request.params.id).attrs;
  });

  actions.post("/user", async (schema: any, request: any) => {
    let attrs = JSON.parse(request.requestBody);
    delete attrs.id;
    if (attrs.nome === "RETORNA_ERRO_208") {
      // resposta Valida porem não esperada.
      return new Response(208, headers, "Erro FAKE");
    }
    if (attrs.nome === "RETORNA_ERRO_400") {
      // resposta Valida porem não esperada.
      return new Response(400, headers, "Erro FAKE 400");
    }
    if (attrs.nome === "RETORNA_ERRO_4XX") {
      // resposta Valida porem não esperada.
      return new Response(499, headers, "Erro FAKE 4XX");
    }
    if (attrs.companyId <= 0) {
      return new Response(417, headers, "Codigo id da Empresa invalido");
    } else {
      // attrs.group = schema.db.groups.find(attrs.groupId);
      // delete attrs.group.access;
      return new Response(201, headers, schema.users.create(attrs));
      // return schema.operators.create(attrs);
    }
  });
  actions.put("/user", (schema: any, request: any) => {
    let attrs = JSON.parse(request.requestBody);
    // attrs.group = schema.db.groups.find(attrs.groupId);
    // return new Response(200, headers, "schema.operators.create(attrs)");
    return new Response(200, headers, schema.users.create(attrs));
    // return schema.operators.create(attrs);
  });
};

/**
 * Valores MOCk para teste da entidade.
 * @type {import("../../User/UserData").IUser[]}
 */
export const listUserMock: IUser[] = [
  {
    id: 1,
    nome: "Administrador Geral",
    email: "root@com.br",
    password: "cm9vdA==", // btoa("root"),
    companyId: 1,
    groupId: listGroupMock[0].id,
    group: { id: listGroupMock[0].id, nome: listGroupMock[0].nome },
  },
  {
    id: 2,
    nome: "Usuário 1",
    email: "adm@com.br",
    password: "YWRt", // btoa("adm"),
    companyId: 1,
    groupId: listGroupMock[1].id,
    group: { id: listGroupMock[1].id, nome: listGroupMock[1].nome },
  },
  {
    id: 3,
    nome: "Usuário",
    email: "teste@com.br",
    password: "dGVzdGU=", // btoa("teste"),
    companyId: 2,
    groupId: listGroupMock[2].id,
    group: { id: listGroupMock[2].id, nome: listGroupMock[2].nome },
  },
];

export const userMock: IUser = {
  id: 4,
  nome: "USUÁRIO TESTE", // campo eleito para realizar teste de erro
  email: "testemock@teste.com.br",
  password: btoa("u5erP@ssword"), // btoa("u5erP@ssword"),
  companyId: 2,
  groupId: listGroupMock[2].id,
  group: { id: listGroupMock[2].id, nome: listGroupMock[2].nome },
};

let confirmPassword = "u5erP@ssword";
// incluído delay pois na localização do erro não funciona se estiver maior que zero
let setDelay = 100;

export let loadUserValidValues = async (userEvent: any, screen: any) => {
  await userEvent.selectOptions(screen.getByLabelText("Grupo"), [
    userMock.groupId.toString(),
  ]);
  await userEvent.type(
    screen.getByLabelText("Senha"),
    atob(userMock.password ? userMock.password : "") //"u5erP@ssword"
  );
  await userEvent.type(
    screen.getByLabelText("Confirmar Senha"),
    confirmPassword
  );
  await userEvent.type(screen.getByLabelText("E-mail"), userMock.email);
  await userEvent.type(screen.getByLabelText("Nome"), userMock.nome, {
    delay: setDelay,
  });
};

export let loadUserInvalidValues = async (userEvent: any, screen: any) => {
  setDelay = 0;
  userMock.id = 0;
  userMock.nome = "AZ";
  userMock.companyId = 0;
  userMock.email = "teste.br";
  userMock.groupId = 0;
  userMock.group = undefined;
  confirmPassword = "32132";
  userMock.password = "Abcdefg8321";
  loadUserValidValues(userEvent, screen);
};

export let loadLoginValidValues = async (userEvent: any, screen: any) => {
  await userEvent.type(
    screen.getByPlaceholderText("E-mail cadastrado"),
    userMock.email
  );
  await userEvent.type(
    screen.getByPlaceholderText("Senha"),
    userMock.password,
    {
      delay: setDelay,
    }
  );
};

export let loadLoginInvalidValues = async (userEvent: any, screen: any) => {
  setDelay = 0;
  userMock.email = "teste.br";
  loadLoginValidValues(userEvent, screen);
};
