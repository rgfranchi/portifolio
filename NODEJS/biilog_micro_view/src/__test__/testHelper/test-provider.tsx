import React from "react";
import { MemoryRouter, Route, Router } from "react-router-dom";

import { objMockSession } from "./test-session";

import { Provider } from "react-redux";

//componentes de tela.
import { ModalLoading } from "../../components/ModalsComponent";

/**
 * Renderiza valores para teste.
 * Utilizando o render do framework @testing-library/react
 * Componente ModalLoading.
 * @param {*} render
 * @param {*} ui
 * @param {*} history
 * @param {*} store
 * @returns
 */
export const renderProviderRouter = (
  render: any,
  ui: any,
  history: any,
  store: any,
  modal: boolean = true
) => {
  // @ts-ignore
  function Wrapper({ children }) {
    return providerRouter(children, history, store, modal);
  }
  // @ts-ignore
  return render(ui, { wrapper: Wrapper });
};

/**
 * ConstrÃ³i com base no renderProviderRouter incluindo MemoryRouter e Router.
 * @param {*} render
 * @param {*} ui
 * @param {*} history
 * @param {*} store
 * @param {*} initialEntries
 * @param {*} path
 * @returns
 */
export const renderProviderMemoryRouter = (
  render: any,
  ui: any,
  history: any,
  store: any,
  initialEntries: any,
  path: string,
  modal: boolean = true
) => {
  const wrapper = (
    <MemoryRouter initialEntries={[initialEntries]}>
      <Route path={path}>{ui}</Route>
    </MemoryRouter>
  );
  return renderProviderRouter(render, wrapper, history, store, modal);
};

/**
 * Cria as funcionalidades do Provider e Router
 * Utilizando o mount do framework Enzyme
 * @param {*} mount
 * @param {*} children
 * @param {*} history
 * @param {*} store
 * @param {boolean} modal
 * @returns
 */
export const mountProviderRouter = async (
  mount: any,
  children: any,
  history: any,
  store: any,
  modal: boolean = true
) => {
  return await mount(providerRouter(children, history, store, modal));
};

export const createSessionCurrentUser = (store: any) => {
  store.dispatch({
    type: "session/currentUser",
    payload: objMockSession,
  });
};

const providerRouter = (
  children: any,
  history: any,
  store: any,
  modal: boolean
) => {
  return (
    <Provider store={store}>
      <Router history={history}>
        {modal && <ModalLoading />}
        {children}
      </Router>
    </Provider>
  );
};
