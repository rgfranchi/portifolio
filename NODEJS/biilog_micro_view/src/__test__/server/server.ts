import { createServer, RestSerializer, Model } from "miragejs";
import { userRoutes } from "./serverUser";
import { listCompanyMock, companyRoutes } from "./serverCompany";
import { listOperatorMock, operatorRoutes } from "./serverOperator";
import { listUserMock } from "./serverUser";
import { listGroupMock, groupRoutes } from "./serverGroup";
import { listForkliftMock, forkliftRoutes } from "./serverForklift";
import { listPlaceMock, placeRoutes } from "./serverPlace";

let isEmpty = false;
// variÃ¡vel para resposta vazia, impede alimentar com array original.
/**
 *
 * @param {boolean} empty
 */
export const forceEmpty = (empty: boolean) => {
  // console.info("ForceEmpty", empty);
  isEmpty = empty;
};

export const getIsEmpty = () => isEmpty;

// https://miragejs.com/docs/main-concepts/database/
// https://miragejs.com/api/classes/db-collection/
// https://miragejs.com/docs/main-concepts/route-handlers/
export const mockServer = () => {
  return createServer({
    serializers: {
      reminder: RestSerializer.extend({
        include: ["list"],
        embed: true,
      }),
    },
    models: {
      // userCurrent: Model,
      // user: Model,
      user: Model,
      company: Model,
      operator: Model,
      group: Model,
      forklift: Model,
      place: Model,
    },
    seeds(server) {
      // server.db.loadData({ userCurrent: isEmpty ? [] : userCurrentMock });
      server.db.loadData({ users: isEmpty ? [] : listUserMock });
      server.db.loadData({ groups: isEmpty ? [] : listGroupMock });
      server.db.loadData({ companies: isEmpty ? [] : listCompanyMock });
      server.db.loadData({ operators: isEmpty ? [] : listOperatorMock });
      server.db.loadData({ forklifts: isEmpty ? [] : listForkliftMock });
      server.db.loadData({ places: isEmpty ? [] : listPlaceMock });
    },
    routes() {
      this.namespace = "api/v0";
      userRoutes(this);
      companyRoutes(this);
      operatorRoutes(this);
      groupRoutes(this);
      forkliftRoutes(this);
      placeRoutes(this);
      this.passthrough("https://viacep.com.br/ws/**");
    },
  });
};
