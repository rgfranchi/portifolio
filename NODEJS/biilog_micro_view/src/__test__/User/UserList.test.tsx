import React from "react";

import { act } from "react-dom/test-utils";
import Enzime, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

import { createMemoryHistory } from "history";

import store from "../../redux/store";

import { mountProviderRouter } from "../testHelper/test-provider";

import UserList from "../../User/UserList";
import { listUserMock } from "../server/serverUser";

Enzime.configure({ adapter: new Adapter() });

let wrapper: any = null;
let history: any = null;
const listMock = listUserMock;

beforeEach(() => {
  history = createMemoryHistory();
});

describe("UserList snapshot", () => {
  test("empty", async () => {
    history.entries[0].key = "EMPTY_01";
    await setWrapperDefault([]);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test("with data", async () => {
    history.entries[0].key = "DATA_01";
    await setWrapperDefault(listMock);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("UserList actions", () => {
  const count = listMock.length;
  const arrPos = 2;

  test("update", async () => {
    await setWrapperDefault(listMock);
    history.entries[0].key = "ACTION_UPDATE";
    const action = wrapper.find("ButtonIconUpdate");
    expect(action.length).toEqual(count);
    action.at(arrPos).simulate("click", { button: 0 });
    expect(history.location.pathname).toBe(
      "/urlFake/update/" + listMock[arrPos].id
    );
  });
  test("delete", async () => {
    const mockFunctionDelete = jest.fn((x) => x);
    await setWrapperDefault(listMock, mockFunctionDelete);
    history.entries[0].key = "ACTION_DEL";
    const action = wrapper.find("ButtonIconDelete");
    expect(action.length).toEqual(count);
    action.at(arrPos).simulate("click", { button: 0 });
    expect(mockFunctionDelete.mock.results[0].value).toBe(listMock[arrPos].id);
  });
});

describe("UserList access", () => {
  const count = listMock.length;

  test("update denied", async () => {
    expect(count).not.toEqual(0);
    await setWrapperDefault(listMock, () => {}, { delete: {} });
    history.entries[0].key = "ACTION_UPDATE";
    const action = wrapper.find("ButtonIconUpdate");
    expect(action.length).toEqual(0);
  });
  test("delete denied", async () => {
    expect(count).not.toEqual(0);
    await setWrapperDefault(listMock, () => {}, { update: {} });
    history.entries[0].key = "ACTION_DEL";
    const action = wrapper.find("ButtonIconDelete");
    expect(action.length).toEqual(0);
  });
});

const setWrapperDefault = async (
  list: any,
  handleDelete: any = () => {},
  access: any = { update: {}, delete: {} }
) => {
  await act(async () => {
    wrapper = await mountProviderRouter(
      mount,
      <UserList
        baseUrl="/urlFake"
        access={access}
        list={list}
        handleDeleteOpen={handleDelete}
      />,
      history,
      store
    );
  });
};

// Enzime.configure({ adapter: new Adapter() });

// test("UserList snapshot empty", () => {
//   const warpper = mount(
//     <UserList
//       baseUrl="/urlFake"
//       access={{ update: {}, delete: {} }}
//       list={[]}
//       handleDeleteOpen={() => {}}
//     />
//   );
//   expect(toJson(warpper)).toMatchSnapshot();
// });

// test("UserList snapshot data", () => {
//   const history = createMemoryHistory();
//   history.entries[0].key = "KEY_01";
//   const warpper = mount(
//     <Router history={history}>
//       <UserList
//         baseUrl="/urlFake"
//         access={{ update: {}, delete: {} }}
//         list={listUserMock}
//         handleDeleteOpen={() => {}}
//       />
//     </Router>
//   );
//   expect(toJson(warpper)).toMatchSnapshot();
// });

// test("UserList click update delete", () => {
//   let history = createMemoryHistory();
//   history.entries[0].key = "KEY_01";
//   const mockFunctionDelete = jest.fn((x) => x);
//   const warpper = mount(
//     <Router history={history}>
//       <UserList
//         baseUrl="/urlFake"
//         access={{ update: {}, delete: {} }}
//         list={listUserMock}
//         handleDeleteOpen={mockFunctionDelete}
//       />
//     </Router>
//   );
//   // testa link update
//   const update = warpper.find("ButtonIconUpdate");
//   expect(update.length).toEqual(3);
//   // verifica link antes.
//   expect(history.location.pathname).toBe("/");
//   // simula click
//   update.at(1).simulate("click", { button: 0 });
//   // verifica link depois.
//   expect(history.location.pathname).toBe("/urlFake/update/2");
//   // testa chamada a função delete
//   const del = warpper.find("ButtonIconDelete");
//   expect(del.length).toEqual(3);
//   // simula click
//   del.at(1).simulate("click");
//   // recebe a posição do elemento no array.
//   expect(mockFunctionDelete.mock.results[0].value).toBe(2);
// });
