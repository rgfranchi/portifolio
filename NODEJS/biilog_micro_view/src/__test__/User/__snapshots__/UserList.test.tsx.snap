// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList snapshot empty 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "EMPTY_01",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "EMPTY_01",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ModalLoading>
      <Modal
        animation={false}
        autoFocus={true}
        backdrop={true}
        centered={true}
        dialogAs={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "ModalDialog",
            "render": [Function],
          }
        }
        enforceFocus={true}
        keyboard={true}
        onHide={[Function]}
        restoreFocus={true}
        show={false}
        size="lg"
      >
        <Modal
          autoFocus={true}
          backdrop={true}
          containerClassName="modal-open"
          enforceFocus={true}
          keyboard={true}
          manager={
            BootstrapModalManager {
              "containers": Array [],
              "data": Array [],
              "handleContainerOverflow": true,
              "hideSiblingNodes": true,
              "modals": Array [],
              "scrollbarSize": 0,
            }
          }
          onEnter={[Function]}
          onEntering={[Function]}
          onEscapeKeyDown={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onHide={[Function]}
          renderBackdrop={[Function]}
          renderDialog={[Function]}
          restoreFocus={true}
          show={false}
        />
      </Modal>
    </ModalLoading>
    <UserList
      access={
        Object {
          "delete": Object {},
          "update": Object {},
        }
      }
      baseUrl="/urlFake"
      handleDeleteOpen={[Function]}
      list={Array []}
    >
      <ForwardRef>
        <table
          className="table"
        >
          <thead>
            <tr>
              <th>
                Nome
              </th>
              <th>
                E-mail
              </th>
              <th>
                Grupo
              </th>
              <th>
                Ações
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </ForwardRef>
    </UserList>
  </Router>
</Provider>
`;

exports[`UserList snapshot with data 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "DATA_01",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "DATA_01",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ModalLoading>
      <Modal
        animation={false}
        autoFocus={true}
        backdrop={true}
        centered={true}
        dialogAs={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "ModalDialog",
            "render": [Function],
          }
        }
        enforceFocus={true}
        keyboard={true}
        onHide={[Function]}
        restoreFocus={true}
        show={false}
        size="lg"
      >
        <Modal
          autoFocus={true}
          backdrop={true}
          containerClassName="modal-open"
          enforceFocus={true}
          keyboard={true}
          manager={
            BootstrapModalManager {
              "containers": Array [],
              "data": Array [],
              "handleContainerOverflow": true,
              "hideSiblingNodes": true,
              "modals": Array [],
              "scrollbarSize": 0,
            }
          }
          onEnter={[Function]}
          onEntering={[Function]}
          onEscapeKeyDown={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onHide={[Function]}
          renderBackdrop={[Function]}
          renderDialog={[Function]}
          restoreFocus={true}
          show={false}
        />
      </Modal>
    </ModalLoading>
    <UserList
      access={
        Object {
          "delete": Object {},
          "update": Object {},
        }
      }
      baseUrl="/urlFake"
      handleDeleteOpen={[Function]}
      list={
        Array [
          Object {
            "companyId": 1,
            "email": "root@com.br",
            "group": Object {
              "id": 1,
              "nome": "Root",
            },
            "groupId": 1,
            "id": 1,
            "nome": "Administrador Geral",
            "password": "cm9vdA==",
          },
          Object {
            "companyId": 1,
            "email": "adm@com.br",
            "group": Object {
              "id": 2,
              "nome": "Administrador",
            },
            "groupId": 2,
            "id": 2,
            "nome": "Usuário 1",
            "password": "YWRt",
          },
          Object {
            "companyId": 2,
            "email": "teste@com.br",
            "group": Object {
              "id": 3,
              "nome": "Usuário Comum",
            },
            "groupId": 3,
            "id": 3,
            "nome": "Usuário",
            "password": "dGVzdGU=",
          },
        ]
      }
    >
      <ForwardRef>
        <table
          className="table"
        >
          <thead>
            <tr>
              <th>
                Nome
              </th>
              <th>
                E-mail
              </th>
              <th>
                Grupo
              </th>
              <th>
                Ações
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              key="1"
            >
              <td>
                Administrador Geral
              </td>
              <td>
                root@com.br
              </td>
              <td>
                Root
              </td>
              <td>
                <ButtonIconUpdate
                  linkTo="/urlFake/update/1"
                >
                  <Button
                    active={false}
                    as={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "Link",
                        "propTypes": Object {
                          "innerRef": [Function],
                          "onClick": [Function],
                          "replace": [Function],
                          "target": [Function],
                          "to": [Function],
                        },
                        "render": [Function],
                      }
                    }
                    disabled={false}
                    size="sm"
                    to="/urlFake/update/1"
                    variant="primary"
                  >
                    <Link
                      className="btn btn-primary btn-sm"
                      disabled={false}
                      to="/urlFake/update/1"
                    >
                      <LinkAnchor
                        className="btn btn-primary btn-sm"
                        disabled={false}
                        href="/urlFake/update/1"
                        navigate={[Function]}
                      >
                        <a
                          className="btn btn-primary btn-sm"
                          disabled={false}
                          href="/urlFake/update/1"
                          onClick={[Function]}
                        >
                          <FiEdit>
                            <IconBase
                              attr={
                                Object {
                                  "fill": "none",
                                  "stroke": "currentColor",
                                  "strokeLinecap": "round",
                                  "strokeLinejoin": "round",
                                  "strokeWidth": "2",
                                  "viewBox": "0 0 24 24",
                                }
                              }
                            >
                              <svg
                                fill="none"
                                height="1em"
                                stroke="currentColor"
                                strokeLinecap="round"
                                strokeLinejoin="round"
                                strokeWidth="2"
                                style={
                                  Object {
                                    "color": undefined,
                                  }
                                }
                                viewBox="0 0 24 24"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                                  key="0"
                                />
                                <path
                                  d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                                  key="1"
                                />
                              </svg>
                            </IconBase>
                          </FiEdit>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </Button>
                </ButtonIconUpdate>
                 
                <ButtonIconDelete
                  execute={[Function]}
                >
                  <Button
                    active={false}
                    disabled={false}
                    onClick={[Function]}
                    size="sm"
                    variant="danger"
                  >
                    <button
                      className="btn btn-danger btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <AiFillDelete>
                        <IconBase
                          attr={
                            Object {
                              "viewBox": "0 0 1024 1024",
                            }
                          }
                        >
                          <svg
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            strokeWidth="0"
                            style={
                              Object {
                                "color": undefined,
                              }
                            }
                            viewBox="0 0 1024 1024"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M864 256H736v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zm-200 0H360v-72h304v72z"
                              key="0"
                            />
                          </svg>
                        </IconBase>
                      </AiFillDelete>
                    </button>
                  </Button>
                </ButtonIconDelete>
              </td>
            </tr>
            <tr
              key="2"
            >
              <td>
                Usuário 1
              </td>
              <td>
                adm@com.br
              </td>
              <td>
                Administrador
              </td>
              <td>
                <ButtonIconUpdate
                  linkTo="/urlFake/update/2"
                >
                  <Button
                    active={false}
                    as={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "Link",
                        "propTypes": Object {
                          "innerRef": [Function],
                          "onClick": [Function],
                          "replace": [Function],
                          "target": [Function],
                          "to": [Function],
                        },
                        "render": [Function],
                      }
                    }
                    disabled={false}
                    size="sm"
                    to="/urlFake/update/2"
                    variant="primary"
                  >
                    <Link
                      className="btn btn-primary btn-sm"
                      disabled={false}
                      to="/urlFake/update/2"
                    >
                      <LinkAnchor
                        className="btn btn-primary btn-sm"
                        disabled={false}
                        href="/urlFake/update/2"
                        navigate={[Function]}
                      >
                        <a
                          className="btn btn-primary btn-sm"
                          disabled={false}
                          href="/urlFake/update/2"
                          onClick={[Function]}
                        >
                          <FiEdit>
                            <IconBase
                              attr={
                                Object {
                                  "fill": "none",
                                  "stroke": "currentColor",
                                  "strokeLinecap": "round",
                                  "strokeLinejoin": "round",
                                  "strokeWidth": "2",
                                  "viewBox": "0 0 24 24",
                                }
                              }
                            >
                              <svg
                                fill="none"
                                height="1em"
                                stroke="currentColor"
                                strokeLinecap="round"
                                strokeLinejoin="round"
                                strokeWidth="2"
                                style={
                                  Object {
                                    "color": undefined,
                                  }
                                }
                                viewBox="0 0 24 24"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                                  key="0"
                                />
                                <path
                                  d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                                  key="1"
                                />
                              </svg>
                            </IconBase>
                          </FiEdit>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </Button>
                </ButtonIconUpdate>
                 
                <ButtonIconDelete
                  execute={[Function]}
                >
                  <Button
                    active={false}
                    disabled={false}
                    onClick={[Function]}
                    size="sm"
                    variant="danger"
                  >
                    <button
                      className="btn btn-danger btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <AiFillDelete>
                        <IconBase
                          attr={
                            Object {
                              "viewBox": "0 0 1024 1024",
                            }
                          }
                        >
                          <svg
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            strokeWidth="0"
                            style={
                              Object {
                                "color": undefined,
                              }
                            }
                            viewBox="0 0 1024 1024"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M864 256H736v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zm-200 0H360v-72h304v72z"
                              key="0"
                            />
                          </svg>
                        </IconBase>
                      </AiFillDelete>
                    </button>
                  </Button>
                </ButtonIconDelete>
              </td>
            </tr>
            <tr
              key="3"
            >
              <td>
                Usuário
              </td>
              <td>
                teste@com.br
              </td>
              <td>
                Usuário Comum
              </td>
              <td>
                <ButtonIconUpdate
                  linkTo="/urlFake/update/3"
                >
                  <Button
                    active={false}
                    as={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "displayName": "Link",
                        "propTypes": Object {
                          "innerRef": [Function],
                          "onClick": [Function],
                          "replace": [Function],
                          "target": [Function],
                          "to": [Function],
                        },
                        "render": [Function],
                      }
                    }
                    disabled={false}
                    size="sm"
                    to="/urlFake/update/3"
                    variant="primary"
                  >
                    <Link
                      className="btn btn-primary btn-sm"
                      disabled={false}
                      to="/urlFake/update/3"
                    >
                      <LinkAnchor
                        className="btn btn-primary btn-sm"
                        disabled={false}
                        href="/urlFake/update/3"
                        navigate={[Function]}
                      >
                        <a
                          className="btn btn-primary btn-sm"
                          disabled={false}
                          href="/urlFake/update/3"
                          onClick={[Function]}
                        >
                          <FiEdit>
                            <IconBase
                              attr={
                                Object {
                                  "fill": "none",
                                  "stroke": "currentColor",
                                  "strokeLinecap": "round",
                                  "strokeLinejoin": "round",
                                  "strokeWidth": "2",
                                  "viewBox": "0 0 24 24",
                                }
                              }
                            >
                              <svg
                                fill="none"
                                height="1em"
                                stroke="currentColor"
                                strokeLinecap="round"
                                strokeLinejoin="round"
                                strokeWidth="2"
                                style={
                                  Object {
                                    "color": undefined,
                                  }
                                }
                                viewBox="0 0 24 24"
                                width="1em"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                                  key="0"
                                />
                                <path
                                  d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                                  key="1"
                                />
                              </svg>
                            </IconBase>
                          </FiEdit>
                        </a>
                      </LinkAnchor>
                    </Link>
                  </Button>
                </ButtonIconUpdate>
                 
                <ButtonIconDelete
                  execute={[Function]}
                >
                  <Button
                    active={false}
                    disabled={false}
                    onClick={[Function]}
                    size="sm"
                    variant="danger"
                  >
                    <button
                      className="btn btn-danger btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <AiFillDelete>
                        <IconBase
                          attr={
                            Object {
                              "viewBox": "0 0 1024 1024",
                            }
                          }
                        >
                          <svg
                            fill="currentColor"
                            height="1em"
                            stroke="currentColor"
                            strokeWidth="0"
                            style={
                              Object {
                                "color": undefined,
                              }
                            }
                            viewBox="0 0 1024 1024"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M864 256H736v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zm-200 0H360v-72h304v72z"
                              key="0"
                            />
                          </svg>
                        </IconBase>
                      </AiFillDelete>
                    </button>
                  </Button>
                </ButtonIconDelete>
              </td>
            </tr>
          </tbody>
        </table>
      </ForwardRef>
    </UserList>
  </Router>
</Provider>
`;
